### PR TITLE
fixes #12763 - puppetclass importer should be async

### PR DESCRIPTION
**Pull request by unorthodoxgeek**

### CODE REVIEW

File: app/lib/actions/foreman/puppetclass/import.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+module Actions
+  module Foreman
+    module Puppetclass
+      class Import < Actions::EntryAction
+        def resource_locks
+          :import_puppetclasses
+        end
+
+        def plan(params)
+          plan_self(params)
```
Comment (iNecas) [inline app/lib/actions/foreman/puppetclass/import.rb]: This is the dafult implementation of the plan method in the action - no need to redefine unless other intention

File: app/controllers/foreman_tasks/concerns/environments_extension.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+      included do
+        alias_method_chain :obsolete_and_new, :dynflow
+      end
+
+      def obsolete_and_new_with_dynflow
+        task = ForemanTasks.async_task(::Actions::Foreman::Puppetclass::Import,
+                                       params)
+
+        notice _("Added import task to queue, it will be run shortly")
```
Comment (iNecas) [inline app/controllers/foreman_tasks/concerns/environments_extension.rb]: A link to the task details would be interesting here.

**Comment by unorthodoxgeek** [comment]
@iNecas - fixed according to comments, please re-review

**Comment by iNecas** [comment]
[test]

File: app/controllers/foreman_tasks/concerns/environments_extension.rb (hunk 2 above)
Comment (iNecas) [inline app/controllers/foreman_tasks/concerns/environments_extension.rb]: I have note realized the link is not shown properly in the notification window, so it just renders in page escaped. Sry about that - we would need a better infra in UI to make it better. So the version is ok.

File: lib/foreman_tasks/engine.rb [hunk 3]
```diff
@@ -69,6 +69,8 @@ class Engine < ::Rails::Engine
 
       config.to_prepare do
         ::Api::V2::HostsController.send :include, ForemanTasks::Concerns::HostsControllerExtension
+        ::PuppetclassesController.send :include, ForemanTasks::Concerns::EnvironmentsExtension
```
Comment (iNecas) [inline lib/foreman_tasks/engine.rb]: Do we need to extend this controller as well? I don't see it defining the `obsolete_and_new`

Comment (unorthodoxgeek) [inline lib/foreman_tasks/engine.rb]: well, technically, the definition is at Foreman::Controller::Environments.

I could include it there, but this seems to be more verbose...

Comment (iNecas) [inline lib/foreman_tasks/engine.rb]: Right, I have misread it completely before (haven't noticed the `Foreman::Controller::Environments` is a concern (I would love to see the modules in Foreman to follow the directory structure…). With this fact, I agree this is better.

File: app/controllers/foreman_tasks/concerns/environments_extension.rb (hunk 2 above)
Comment (unorthodoxgeek) [inline app/controllers/foreman_tasks/concerns/environments_extension.rb]: reverted the reverted reverted code.

Comment (iNecas) [inline app/controllers/foreman_tasks/concerns/environments_extension.rb]: :ok_hand: :)

**Comment by iNecas** [comment]
Thanks @unorthodoxgeek

**Pull request merged**